### PR TITLE
Support runtimeconfig.dev.json

### PR DESCRIPTION
**Pull request by schellap**
WIP because of testing.

@gkhanna79

---
This change is [Reviewable](https://reviewable.io/reviews/dotnet/cli/2191)

### CODE REVIEW

**Comment by davidfowl** [comment]
Isn't it `dev.runtimeconfig.json`?

/cc @livarcocc

**Comment by gkhanna79** [comment]
@schellap Errors like the one below:

08:12:03 #pragma clang diagnostic ignored "-Wunused-local-typedef"
08:12:03                                  ^
08:12:03 In file included from /mnt/resource/j/workspace/dotnet_cli/rel_1.0.0/debug_centos7.1_x64_prtest/src/corehost/cli/runtime_config.cpp:8:
08:12:03 /mnt/resource/j/workspace/dotnet_cli/rel_1.0.0/debug_centos7.1_x64_prtest/src/corehost/cli/./runtime_config.h:19:16: error: no type named 'list' in namespace 'std'
08:12:03     const std::list<pal::string_t>& get_probe_paths() const;

**Comment by gkhanna79** [comment]
@davidfowl It can be either - is there a specific concern with the name?

File: src/corehost/cli/fxr/fx_muxer.cpp [hunk 1]
```diff
@@ -257,24 +257,26 @@ int fx_muxer_t::execute(const int argc, const pal::char_t* argv[])
             return StatusCode::LibHostExecModeFailure;
         }
 
-        auto config_file = get_runtime_config_from_file(app_path);
-        runtime_config_t config(config_file);
+        pal::string_t dev_config_file;
+        auto config_file = get_runtime_config_from_file(app_path, &dev_config_file);
+        runtime_config_t config(config_file, dev_config_file);
         if (!config.is_valid())
```
Comment (gkhanna79) [inline src/corehost/cli/fxr/fx_muxer.cpp]: How do you handle the case of runtimeconfig.dev.json being invalid or absent?

File: src/corehost/cli/fxr/fx_muxer.cpp [hunk 2]
```diff
@@ -307,18 +309,24 @@ int fx_muxer_t::execute(const int argc, const pal::char_t* argv[])
...
+            runtime_config_t config(config_file, dev_config_file);
+            for (const auto& path : config.get_probe_paths())
+            {
+                probe_paths.push_back(path);
+            }
+
             if (!config.is_valid())
             {
-                trace::error(_X("Invalid runtimeconfig.json [%s]"), config.get_path().c_str());
+                trace::error(_X("Invalid runtimeconfig.json [%s] [%s]"), config.get_path().c_str(), config.get_dev_path().c_str());
```
Comment (gkhanna79) [inline src/corehost/cli/fxr/fx_muxer.cpp]: Same as earlier.

File: src/corehost/cli/fxr/fx_muxer.cpp [hunk 3]
```diff
@@ -338,14 +346,25 @@ int fx_muxer_t::execute(const int argc, const pal::char_t* argv[])
             {
                 trace::verbose(_X("Executing as a standalone app as per config file [%s]"), config_file.c_str());
                 pal::string_t impl_dir = get_directory(app_or_deps);
-                if (!library_exists_in_dir(impl_dir, LIBHOSTPOLICY_NAME, nullptr) && !probe_path.empty() && !deps_file.empty())
+                if (!library_exists_in_dir(impl_dir, LIBHOSTPOLICY_NAME, nullptr) && !probe_paths.empty() && !deps_file.empty())
```
Comment (gkhanna79) [inline src/corehost/cli/fxr/fx_muxer.cpp]: Isnt this the case of resolving hostpolicy from dev nuget cache? If so, then we should only be consulting the probe path for that cache and not the entire set.

File: src/corehost/cli/libhost.cpp [hunk 4]
```diff
@@ -6,14 +6,19 @@
 #include "trace.h"
 #include "libhost.h"
 
-pal::string_t get_runtime_config_from_file(const pal::string_t& file)
+pal::string_t get_runtime_config_from_file(const pal::string_t& file, pal::string_t* dev_cfg)
```
Comment (gkhanna79) [inline src/corehost/cli/libhost.cpp]: Why do we need to pass dev_cfg if it not used in every callsite?

Comment (gkhanna79) [inline src/corehost/cli/libhost.cpp]: Nevermind - got it.

**Comment by schellap** [comment]
Todo: Set roll fwd to false if dev.json present even though it doesn't have a field.

**Comment by gkhanna79** [comment]
> Todo: Set roll fwd to false if dev.json present even though it doesn't have a field.

This is required for RC2.

**Comment by schellap** [comment]
yes, just calling out because I remembered it in the shower :)

**Comment by livarcocc** [comment]
@davidfowl I followed the same pattern as app.runtimeconfig.json and made it app.runtimeconfig.dev.json. So, if app name is example, it will be example.runtimeconfig.dev.json.

**Comment by davidfowl** [comment]
Ok

**Comment by schellap** [comment]
@gkhanna79 passes my validations by hand. @Sridhar-MS can confirm with my bin drop. Waiting for your :+1:  after an eyeball pass.

**Comment by schellap** [comment]
@Sridhar-MS just confirmed MusicStore is good on this PR's bin drop. But in any case, a CR would be good to have.

**Comment by schellap** [comment]
@dotnet-bot test Windows_NT x86 Debug Build please

**Comment by gkhanna79** [comment]
Awesome!

**Pull request merged**